design/20230302.gomod.md

```diff
@@ -0,0 +1,250 @@
+
+Between Helm 3.10 and 3.11, several of Helm's dependencies were upgraded, and crucially Helm has some of the same
+dependencies that cert-manager does. That means that we can't easily _just_ upgrade Helm.
+
+Running `go get -u helm.sh/helm/v3` produces 56 different upgrades of _other_ dependencies. Notably, it bumps our
+Kubernetes dependencies from v0.25.2 to v0.26.0 but there are several other changes.
+
+(NB: Helm is just an example here and we could have problems with any package)
+
+## Proposed Solution: Go Module Proliferation
+
+**In short:** Add several new `go.mod` files so individual components can be patched independently
+
+We can create several new Go modules so that each binary we build can have distinct dependencies. This would mean that
+`cmctl` having a dependency on Helm would only affect `cmctl` and wouldn't force us to change any of the other
+components we build in order to patch a Helm vulnerability.
+
+In addition to separating out the binaries, we could also separate out the "core" issuer implementations into a
+distinct module (since they bring so many external dependencies and currently live in `pkg/` where they'll be part of
+the core cert-manager `go.mod` file).
+
+Plus, where we have testing-only dependencies (e.g. for integration or end-to-end tests) we could create a test module
+so that those test dependencies don't pollute the main `go.mod`.
+
+## Potential Issues
+
+### Importing cert-manager / Development Experience
+
+**In short:** Module replacements help developers but don't aren't respected by imports, meaning some changes could need two PRs or we'd have to break importing certain modules
+
+**Useful Reference:** It helps to read [this StackOverflow comment](https://stackoverflow.com/a/71984158) to better understand the options we have
+
+The simplest development experience when working with multiple Go modules at once is to use either the `replace` directive
+in `go.mod` or the `use` directive in `go.work` to point to local versions of a module. This allows both modules to be
+developed in parallel on a local machine.
+
+For modules which we don't think should ever be imported by third parties, replace directives would work so that those
+modules always use the version of cert-manager which is at the same commit as those modules.
+
+For example we could look at the `controller` component which would depend on the core cert-manager module. Its
+`go.mod` might contain the following:
+
+```text
+module github.com/cert-manager/cert-manager/cmd/controller
+
+...
+
+require (
+	github.com/cert-manager/cert-manager v0.0.0-00010101000000-000000000000
+	...
+)
+
+...
+
+replace github.com/cert-manager/cert-manager => ../../
+
+...
+```
+
+An issue with this approach is that the `replace` statement wouldn't be respected if anyone imports
+`github.com/cert-manager/cert-manager/cmd/controller` from a 3rd party project. Instead, that 3rd party would see an
+error relating to an unknown version of cert-manager.
```

Comment:
That's a potential solution, but I worry that it means the code we test against for our releases will differ from what third parties will actually run when importing. If we have the two-PR system, then every single PR will be fully tested so that the `master` branch is always fully tested.